### PR TITLE
Add support for "Scoped" lifetime

### PR DESCRIPTION
**Pull request by jazzman55**
## Description

- Introduced _CreateScope_ method to _IServiceProvider_.
- _ServiceProviderScope_ class, which has its own _IServiceCollection_ of _scoped_ services.
- Modifications to _ServiceProviderEngine_ to look for _scoped_ service in the collection passed from _ServiceProviderScope_
- New unit tests in _DependencyInjectionTests_

## Motivation and Context

- Resolves nanoFramework/Home#1353

## How Has This Been Tested?

New and existing unit tests.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

File: nanoFramework.DependencyInjection/DependencyInjection/ServiceProviderScope.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// Copyright (c) .NET Foundation and Contributors
+// See LICENSE file in the project root for full license information.
+//
+
+using System;
+
+namespace nanoFramework.DependencyInjection
+{
+    /// <summary>
+    /// Default implementation if <see cref="IServiceProviderScope"/>.
```
Comment (josesimoes) [inline nanoFramework.DependencyInjection/DependencyInjection/ServiceProviderScope.cs:11]: Typo maybe...
```suggestion
    /// Default implementation of <see cref="IServiceProviderScope"/>.
```

File: nanoFramework.DependencyInjection/System/IServiceProvider.cs [hunk 2]
```diff
@@ -27,5 +27,11 @@ public interface IServiceProvider
...
         /// </returns>
         object[] GetService(Type[] serviceType);
+
+        /// <summary>
+        /// Creates a scope for resolving services with Scoped lifetime.
+        /// </summary>
+        /// <returns>A service provider scope.</returns>
+        IServiceProviderScope CreateScope();
     }
 }
```
Comment (josesimoes) [inline nanoFramework.DependencyInjection/System/IServiceProvider.cs:37]:
```suggestion
}

```

File: nanoFramework.DependencyInjection/System/IServiceProviderScope.cs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace System
+{
+    /// <summary>
+    /// Defines scope for <see cref="IServiceProvider"/>.
+    /// </summary>
+    public interface IServiceProviderScope : IServiceProvider, IDisposable
+    {
+    }
+}
```
Comment (josesimoes) [inline nanoFramework.DependencyInjection/System/IServiceProviderScope.cs:14]:
```suggestion
}

```

**Review by josesimoes — COMMENTED** [review]
Nice improvement! Thanks.

**Comment by josesimoes** [comment]
@jazzman55 please follow the instructions on the PR template: do not add references or mentions to IDs or similar in the title.
This is already tracked in the PR comment.
Release notes are created from the PR title and there should be only one sentence capturing the PR. 😉

**Comment by josesimoes** [comment]
@jazzman55 seems that there are merge conflicts when updating from upstream. 
Can you please rebase manually and commit that?

**Review by bytewizer — COMMENTED** [review]
@jazzman55 -  Overall this feature looks great.  Nice work! However, I do have a few suggestions:

1.  To keep things consistent with dotnet core source code naming we should be using the  dotnet core class naming.  The usage interface looks good but we should be using the following class naming that it is constant with the full dotnet core.

     ServiceProviderScope => ServiceProviderEngineScope
     IServiceProviderScope => IServiceScope
     IServiceScopeFactory

2. In the spirit keeping things consistent we should add ValidateScopes option to the ServiceProviderOptions and then enable/disable validating the scopes.

```csharp
/// <summary>
/// <c>true</c> to perform check verifying that scoped services never gets resolved from root provider; otherwise <c>false</c>. Defaults to <c>false</c>.
/// </summary>
public bool ValidateScopes { get; set; }
```

3. Please try and match the dotnet source naming, comments documentation and structure as much as possible.  I know this this is a "mini" version but my goal from the start has been to make them match as much as possible.
 
https://github.com/dotnet/runtime/tree/main/src/libraries

**Comment by jazzman55** [comment]
@bytewizer regarding your comments:
1. Renamed. I skipped IServiceScopeFactory as we don't need it now.
2. done
3. Done. Makes great sense to keep things in line with the "big" .NET

**Comment by josesimoes** [comment]
@jazzman55 thank you for going the extra mile and improve this PR even further!👍
And yes, we try as much as possible to align with the full .NET. 😅

**Comment by microcompiler** [comment]
@jazzman55 very nice job!  Overall, it looks good.   I would like to take a couple days to pull down the code and compile it myself.   I don’t anticipate any major changes but would like to kick the tires on it myself before approving it.

**Comment by jazzman55** [comment]
@microcompiler cool! Feel free to reach me out on discord if you need anything.

**Comment by josesimoes** [comment]
@microcompiler @bytewizer when do you think you can put some time to review this? 😁

**Comment by bytewizer** [comment]
> @microcompiler @bytewizer when do you think you can put some time to review this? 😁

Sorry folks!   This has taken me longer than I had hoped for.   I'm almost done and just working right now on the documentation and adding a sample application.     I don't use scoped features very often and took me quite a bit more time than I would have liked trying things both in regular dotnet and nano to make sure the behavior was exact.   Overall, the changes are very solid.  I did make a few changes to the scope engine.  I'll be making a commit in the next day or two for everyone's review.

**Comment by bytewizer** [comment]
@jazzman55 @josesimoes Please have a look at commit https://github.com/nanoframework/nanoFramework.DependencyInjection/commit/4bd22ef8f204fddc53a381d8592f54bf5756c64e   for my recommend changes

**Comment by josesimoes** [comment]
> @jazzman55 @josesimoes Please have a look at commit [4bd22ef](https://github.com/nanoframework/nanoFramework.DependencyInjection/commit/4bd22ef8f204fddc53a381d8592f54bf5756c64e) for my recommend changes

OK. Thank you for the review and improvements.
I've submitted PR CShark-Hub/nanoFramework.DependencyInjection#1 against the origin branch of this PR in order to get your changes into it.
@jazzman55 please review (and possibly merge) when you have the opportunity.

**Comment by josesimoes** [comment]
@jazzman55 as you are at it, please merge from `main` so we have a build with everything up to date.

**Comment by jazzman55** [comment]
Hey @bytewizer , can you look at the sonar bug? I added unit test for it: _IServiceProviderCreateScopeReturnsNewScope_ in [CSHARK-Hub's branch](https://github.com/CShark-Hub/nanoFramework.DependencyInjection/tree/scopes)

**Comment by bytewizer** [comment]
@ jazzman55 - This https://github.com/CShark-Hub/nanoFramework.DependencyInjection/pull/2/commits/ec02457d212f6ca1f89f665376599a2a863f006b fix the sonar bug issue.

**Review by bytewizer — APPROVED** [review]
Looks Good!

**Comment by josesimoes** [comment]
@jazzman55 @bytewizer thank you guys for your work on this PR! Great team work here. 👍🏻 👏🏻 😃

**Pull request merged**